index.ts

```diff
@@ -186,19 +194,119 @@ const llamaInvoke = (prompt: string, input: string): Promise<string> => {
+    let promisesChain = Promise.resolve();
+
+    const sentences: string[] = [];
+    let currentSentence: string[] = [];
+    llama.startAsync({
+      prompt: formattedPrompt,
+      listener: (token: string) => {
+        if (token === '[end of text]') {
+          // Add any remaining tokens to the sentences array.
+          if (currentSentence.length > 0) {
+            sentences.push(currentSentence.join(''));
+          }
+          return resolve(promisesChain.then(() => sentences.join(' ')));
+        }
+
+        if (promptTokensDoneEchoing) {
+          token = token.replace(/[^a-zA-Z0-9 .,!?'\n-]/g, '');
+          currentAudioGeneration= sentences.flatMap((t) => t).join('');
+          updateScreen();
+          stdout.write(currentSentence.join(''))
+          currentSentence.push(token);
+          // Check if the token ends a sentence.
+          if (sentenceEndRegex.test(token)) {
+            const sentence = currentSentence.join('');
+            sentences.push(sentence);
+            const promise = generateAudio(sentence);
+            promisesChain = promisesChain.then(() => promise.then(playAudioFile));
+            sentenceEndRegex = /[.!?]/;  // less greedy
+            currentSentence = [];
+          }
+        } else {
+          promptTokens.push(token)
+        }
+        // TODO: The trim is leak from llama tokenizer inside the cpp
+        // Ideally the cpp binding can take an option on whether to echo prompt tokens or not 
+        // so I don't have to leak the lurk in here
+        if (!promptTokensDoneEchoing && promptTokens.join('').trim() === formattedPrompt.trim()) {
+          promptTokensDoneEchoing = true;
+        }
+      }
+    });
+  });
+}
+
+const getLlamaInputData = (conversation: Conversation): string => {
+  // unclear behavior change
+  const transcriptionSinceWatermark = getTranscriptionSinceWatermark(conversation);
+  const userSpeech: Speech = { speaker: "user", response: transcriptionSinceWatermark };
+  return [...conversation.canonicalDialogue.slice(-4), userSpeech]
+    .map((speech) => `${speech.speaker}: ${speech.response}`)
+    .join('\n');
+}
+
 const llamaLoop = async () => {
   await new Promise(r => setTimeout(r, 2000));
   while (true) {
-    await new Promise(r => setTimeout(r, 30));
+    await new Promise(r => setTimeout(r, 1000));
     if (conversation.transcriptions.length > 0) {
       const currentTranscription = conversation.transcriptions[conversation.transcriptions.length - 1];
-      globalLlamaPromise = llamaInvoke("Be extremely terse. Given the following transcript, respond to it as if there was a conversation.", 'transcript:' + currentTranscription.text);
+      const inputData = getLlamaInputData(conversation);
+      globalLlamaPromise = llamaInvoke("Be extremely terse. Come up with a preamble for the next diagloe. Not the entire response! Just a reasonable 3 word sentence to start with", inputData);
```

Comment:
```suggestion
      globalLlamaPromise = llamaInvoke("Be extremely terse. Come up with a preamble for the next dialogue. Not the entire response! Just a reasonable 3 word sentence to start with", inputData);
 ```
 
 Correct spelling helps out the tokenizer a lot :)

Reply:
Ah!